### DISCUSSION TITLE
Migrating to remix vite:dev cloudflare fails to create cookie, crypto is not defined.

### DISCUSSION THREAD

**Discussion by KeiranHines** [remix-run/remix#8883] [Q&A] [answered]
I get the following error when trying to create a session cookie with cloudflare on remix:vite version 2.7.2

```
ReferenceError: crypto is not defined
    at createKey (<path>\node_modules\@remix-run\cloudflare\dist\crypto.js:34:13)
    at unsign (<path>\node_modules\@remix-run\cloudflare\dist\crypto.js:27:19)
    at decodeCookieValue (<path>\node_modules\@remix-run\server-runtime\dist\cookies.js:92:33)
    at Object.parse (<path>\node_modules\@remix-run\server-runtime\dist\cookies.js:64:66)
    at Object.getSession (<path>\node_modules\@remix-run\server-runtime\dist\sessions.js:118:46)
    at getSession (<path>\app\session.server.ts:76:33)
```

Any help to fix would be appreciated. I tried adding the installGlobals polyfill but that did not work. 

Thanks, 
Keiran.

**Comment by sergiodxa** [comment, marked as answer]
What version of Node are you using? In v20 crypto is global like in CF so it may work.

**Reply by KeiranHines** [reply]
was using 18.17.1, I checked against the cloudflare template `npx create-remix@latest --template remix-run/remix/templates/vite-cloudflare` which specifies >=18.0.0 so I thought that would have been fine. Updating to 20.11.1 seems to have fixed at least that issue, one issue down, others to go. 
 
 Could I suggest either a docs update to mention this, or updating the node version version in that template to 20+ to prevent issues like this in the future for other people.

**Reply by zwily** [reply]
Using Node 20 works for dev, but then when I try to deploy, the wrangler build step fails with

```
✘ [ERROR] Could not resolve "crypto"

    ../node_modules/cookie-signature/index.js:5:21:
      5 │ var crypto = require('crypto');
        ╵                      ~~~~~~~~

  The package "crypto" wasn't found on the file system but is built into node.
  Add the "nodejs_compat" compatibility flag to your Pages project and make sure to prefix the module name with "node:" to enable Node.js compatibility.
```

**Reply by zwily** [reply]
I take that back. Looks like that happens when trying to deploy a vite remix app with any session storage. :/ (tried node 18 and node 20)

**Reply by zwily** [reply]
Wait, my bad. I was accidentally importing `@remix-run/node` somewhere in my app. I will leave this here for my shame to be seen by all. :)

**Reply by kiliman** [reply]
There's an RFC to migrate everything back to a global "remix" package that has all of the shared functionality. You'll only need to import from platform packages for specific functions. This should eliminate issues like this.

Use the remix package RFC https://github.com/remix-run/remix/discussions/7823

**Reply by wallyfaye** [reply]
Updating the node engine in the package.json to version 20 and above fixed this issue for me (I also had to create a `.npmrc` file with `engine-strict=true` in it)

**Reply by anyuruf** [reply]
@Zwily I was using @remix-run/node in places too. I have that cleared but still had to add the compatibilty_flag for kinda a similar error after the crypto one cleared.

**Reply by jordanful** [reply]
Thanks y'all. I had to change my imports from `@remix-run/node` to `@remix-run/cloudflare` everywhere in my app, except `vite.config.ts`, where I call `installGlobals()` imported from `@remix-run/node`. 😵‍💫
